File: src/utils/accessControl.ts
import { Building, BuildingMetadata, Room } from '../core/interfaces/types/buildings';
import { Door } from '../core/interfaces/types/building';

/**
 * Check if a character has access to a building based on metadata permissions.
 */
export function canCharacterAccessBuilding(
    characterId: string,
    metadata?: BuildingMetadata
): boolean {
    if (!metadata) return false;
    if (metadata.owner === characterId) return true;
    if (metadata.residents?.includes(characterId)) return true;
    if (metadata.accessPermissions?.[characterId]) {
        const perm = metadata.accessPermissions[characterId];
        if (!perm.expiresAt || perm.expiresAt > Date.now()) return true;
    }
    // Check for valid key
    if (metadata.keys) {
        for (const keyId in metadata.keys) {
            const key = metadata.keys[keyId];
            if (key.holderId === characterId && (!key.expiresAt || key.expiresAt > Date.now())) {
                return true;
            }
        }
    }
    return false;
}

/**
 * Check if a character has access to a room (room-level permissions override building-level).
 */
export function canCharacterAccessRoom(
    characterId: string,
    room: Room,
    buildingMetadata?: BuildingMetadata
): boolean {
    if (room.accessPermissions?.[characterId]) {
        const perm = room.accessPermissions[characterId];
        if (!perm.expiresAt || perm.expiresAt > Date.now()) return true;
    }
    if (room.residents?.includes(characterId)) return true;
    // Fallback to building-level
    return canCharacterAccessBuilding(characterId, buildingMetadata);
}

/**
 * Check if a character can open a door (door-level permissions override room/building-level).
 */
export function canCharacterAccessDoor(
    characterId: string,
    door: Door,
    room?: Room,
    buildingMetadata?: BuildingMetadata
): boolean {
    if (door.accessPermissions?.[characterId]) {
        const perm = door.accessPermissions[characterId];
        if (!perm.expiresAt || perm.expiresAt > Date.now()) return true;
    }
    // If door is locked and requires a key, check if character has the key
    if (door.isLocked && door.requiredKey && buildingMetadata?.keys) {
        const key = buildingMetadata.keys[door.requiredKey];
        if (key && key.holderId === characterId && (!key.expiresAt || key.expiresAt > Date.now())) {
            return true;
        }
    }
    // Fallback to room/building-level
    if (room && canCharacterAccessRoom(characterId, room, buildingMetadata)) return true;
    return false;
} 

File: backend/utils/access_control.py
from typing import Optional, Dict, Any
import time

def can_character_access_building(character_id: str, metadata: Optional[Dict[str, Any]] = None) -> bool:
    if not metadata:
        return False
    if metadata.get('owner') == character_id:
        return True
    if character_id in metadata.get('residents', []):
        return True
    access_permissions = metadata.get('accessPermissions', {})
    if character_id in access_permissions:
        perm = access_permissions[character_id]
        if not perm.get('expiresAt') or perm['expiresAt'] > int(time.time() * 1000):
            return True
    # Check for valid key
    keys = metadata.get('keys', {})
    for key in keys.values():
        if key.get('holderId') == character_id and (not key.get('expiresAt') or key['expiresAt'] > int(time.time() * 1000)):
            return True
    return False

def can_character_access_room(character_id: str, room: Dict[str, Any], building_metadata: Optional[Dict[str, Any]] = None) -> bool:
    access_permissions = room.get('accessPermissions', {})
    if character_id in access_permissions:
        perm = access_permissions[character_id]
        if not perm.get('expiresAt') or perm['expiresAt'] > int(time.time() * 1000):
            return True
    if character_id in room.get('residents', []):
        return True
    # Fallback to building-level
    return can_character_access_building(character_id, building_metadata)

def can_character_access_door(character_id: str, door: Dict[str, Any], room: Optional[Dict[str, Any]] = None, building_metadata: Optional[Dict[str, Any]] = None) -> bool:
    access_permissions = door.get('accessPermissions', {})
    if character_id in access_permissions:
        perm = access_permissions[character_id]
        if not perm.get('expiresAt') or perm['expiresAt'] > int(time.time() * 1000):
            return True
    # If door is locked and requires a key, check if character has the key
    if door.get('isLocked') and door.get('requiredKey') and building_metadata and 'keys' in building_metadata:
        key = building_metadata['keys'].get(door['requiredKey'])
        if key and key.get('holderId') == character_id and (not key.get('expiresAt') or key['expiresAt'] > int(time.time() * 1000)):
            return True
    # Fallback to room/building-level
    if room and can_character_access_room(character_id, room, building_metadata):
        return True
    return False 